src/shared/components/home/admin-settings.tsx

```diff
@@ -490,4 +519,93 @@ export class AdminSettings extends Component<
     snapToTop();
     await this.fetchUploadsOnly();
   }
+
+  async handleEditOAuthProvider(form: EditOAuthProvider) {
+    this.setState({ loading: true });
+
+    const res = await HttpService.client.editOAuthProvider(form);
+
+    if (res.state === "success") {
+      const newOAuthProvider = res.data;
+      this.setState(s => {
+        s.siteRes.admin_oauth_providers = (
+          s.siteRes.admin_oauth_providers || []
+        ).map(p => {
+          return p?.client_id === newOAuthProvider.client_id
+            ? newOAuthProvider
+            : p;
+        });
+        return s;
+      });
+      toast(I18NextService.i18n.t("site_saved"));
+    } else {
+      toast(I18NextService.i18n.t("couldnt_edit_oauth_provider"), "danger");
+    }
+
+    this.setState({ loading: false });
+
+    return res;
+  }
+
+  async handleDeleteOAuthProvider(form: DeleteOAuthProvider) {
+    this.setState({ loading: true });
+
+    const res = await HttpService.client.deleteOAuthProvider(form);
+
+    let succeeded = false;
+    if (res.state === "success") {
+      this.setState(s => {
+        s.siteRes.admin_oauth_providers = (
+          s.siteRes.admin_oauth_providers || []
+        ).filter(p => p?.id !== form.id);
+        return s;
+      });
+      toast(I18NextService.i18n.t("site_saved"));
+      succeeded = true;
+    } else {
+      toast(I18NextService.i18n.t("couldnt_delete_oauth_provider"), "danger");
+    }
+
+    this.setState({ loading: false });
+
+    return succeeded;
+  }
+
+  async handleCreateOAuthProvider(
+    form: CreateOAuthProvider,
+  ): Promise<OAuthProvider | null> {
+    this.setState({ loading: true });
+
+    const res = await HttpService.client.createOAuthProvider(form);
+    let newOAuthProvider: OAuthProvider;
+
+    if (res.state === "success") {
+      newOAuthProvider = res.data;
+
+      this.setState(s => {
+        s.siteRes.admin_oauth_providers = (
+          s.siteRes.admin_oauth_providers || []
+        ).slice();
+        const index = s.siteRes.admin_oauth_providers.findIndex(
+          x => x?.id === newOAuthProvider?.id,
+        );
+        if (index >= 0) {
+          Object.assign(
+            s.siteRes.admin_oauth_providers[index] || {},
```

Comment:
I greatly doubt this Object.assign is necessary. Check out the `editListImmutable` function we have.